### PR TITLE
IC-734 - Change `ServiceCategory` structure and endpoint

### PR DESCRIPTION
**Pull request by Gweaton**
## What does this pull request do?

Updates service category structure and endpoint so we're no longer storing parts of it on the referral, other than its ID.

## What is the intent behind these changes?

We previously had a `serviceCategory` object on the `DraftReferral` that
had its own ID, and then would make a separate request for the
`complexityLevels` for that `serviceCategory`. Now we're adding more
fields to the `serviceCategory`, it feels as though we might want it to
be a more fully-fledged object, and to store a reference to its `id` on
the `Referral` instead.

Now, a request to `service-category/:id` will return:
```
  {
    id: 'uuid',
    name: 'accommodation',
    complexityLevels: [
     {
       'id': 'uuid',
       'title': 'low complexity',
       'description': 'complexity level description',
     }
    ]
  }
```

The `referral` has a `serviceCategoryId` instead of the whole service
category.

This does mean we're making more requests to the API for both entities,
so we might want to consider the performance implications of this.

For further context, see:
https://github.com/ministryofjustice/hmpps-interventions-ui/pull/44#discussion_r543222831

### CODE REVIEW

**Comment by tomsmyers** [comment]
this looks good to me.

File: server/routes/referrals/completionDeadlinePresenter.test.ts [hunk 1]
```diff
@@ -4,15 +4,19 @@ describe('CompletionDeadlinePresenter', () => {
...
+            serviceCategoryId: 'b33c19d1-7414-4014-b543-e543e59c5b39',
+            complexityLevelId: null,
+          },
+          {
             id: 'b33c19d1-7414-4014-b543-e543e59c5b39',
             name: 'social inclusion',
-          },
-          complexityLevelId: null,
-        })
+            complexityLevels: [],
```
Comment (lawrence-forooghian) [inline server/routes/referrals/completionDeadlinePresenter.test.ts:17]: Now that #45 is merged, could we create a factory for creating `ServiceCategory` instances, and update the tests to use that? (Easy way to find out all the places you need to change: add an extra dummy property to `ServiceCategory` and see all the places the compiler throws an error.)

File: server/routes/referrals/completionDeadlinePresenter.ts [hunk 2]
```diff
@@ -15,17 +15,20 @@ export default class CompletionDeadlinePresenter {
...
 
   constructor(
     private readonly referral: DraftReferral,
+    private readonly serviceCategory: ServiceCategory,
     errors: CompletionDeadlineErrors | null = null,
     userInputData: Record<string, unknown> | null = null
   ) {
     if (!userInputData) {
-      const calendarDay = referral.completionDeadline ? CalendarDay.parseIso8601(referral.completionDeadline) : null
+      const calendarDay = this.referral.completionDeadline
```
Comment (lawrence-forooghian) [inline server/routes/referrals/completionDeadlinePresenter.ts:29]: This looks unrelated to the rest of the changes, was it intentional?

**Review by lawrence-forooghian — CHANGES_REQUESTED** [review]

File: server/routes/referrals/completionDeadlinePresenter.ts (hunk 2 above)
Comment (Gweaton) [reply server/routes/referrals/completionDeadlinePresenter.ts:29]: Yeah, although maybe it should have been in its own commit. I may be misremembering how Typscript instance variables work, but how did this work before? My editor highlighted `referral.completionDeadline` as an error, how come it wasn't `this.referral.completionDeadline` and did that not throw errors for you?

Comment (lawrence-forooghian) [reply server/routes/referrals/completionDeadlinePresenter.ts:29]: I think it worked because `referral` is also a parameter of the constructor so it's available inside the constructor as a local variable, as any other method parameter would be? That's my guess.

Comment (Gweaton) [reply server/routes/referrals/completionDeadlinePresenter.ts]: Right, well now my editor isn't complaining about when I put it back, so have undone this (although prettier still split it onto multiple lines)

File: server/routes/referrals/completionDeadlinePresenter.test.ts (hunk 1 above)
Comment (Gweaton) [reply server/routes/referrals/completionDeadlinePresenter.test.ts]: Done 😓

Comment (Gweaton) [reply server/routes/referrals/completionDeadlinePresenter.test.ts]: Although as I rebased, it became difficult to commit smaller things in this PR (i.e. changes to the `ServiceCategoryFactory`, which I would have liked to be in their own commit), but I've spent so long merging and rebasing other branches over the past few days I don't really have the patience to sort this out...

File: server/routes/referrals/completionDeadlinePresenter.test.ts [hunk 3]
```diff
@@ -72,19 +83,19 @@ describe('CompletionDeadlinePresenter', () => {
...
     it('returns a title', () => {
-      const referral = draftReferralFactory
-        .serviceCategorySelected()
-        .build({ serviceCategory: { name: 'social inclusion' } })
-      const presenter = new CompletionDeadlinePresenter(referral)
+      const serviceCategory = serviceCategoryFactory.build()
+      const referral = draftReferralFactory.serviceCategorySelected(serviceCategory.id).build()
+      const presenter = new CompletionDeadlinePresenter(referral, serviceCategory)
 
-      expect(presenter.title).toEqual('What date does the social inclusion service need to be completed by?')
```
Comment (lawrence-forooghian) [inline server/routes/referrals/completionDeadlinePresenter.test.ts:80]: I think that it's good practice not to rely on the values which are set in the factory. The idea of the factory, as I understand it, is to provide you with a valid, realistic object, possibly in a predefined state. But the actual values inside are subject to change. So I think it's better to write this as

```typescript
const serviceCategory = serviceCategoryFactory.build(name: "social inclusion")
```

and keep the assertion as it was.

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 4]
```diff
@@ -91,9 +69,9 @@ describe('ComplexityLevelPresenter', () => {
 
   describe('title', () => {
     it('returns a title', () => {
-      const presenter = new ComplexityLevelPresenter(draftReferral, socialInclusionComplexityLevels)
+      const presenter = new ComplexityLevelPresenter(draftReferral, serviceCategory)
 
-      expect(presenter.title).toEqual('What is the complexity level for the social inclusion service?')
+      expect(presenter.title).toEqual('What is the complexity level for the accommodation service?')
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:74]: Same as other comment - I think better not to rely on the values in the factory, where possible, and to explicitly set the name of the service category when creating it.

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 5]
```diff
@@ -61,15 +39,15 @@ describe('ComplexityLevelPresenter', () => {
     describe('when the referral already has a selected complexity level', () => {
       it('sets checked to true for the referral’s selected complexity level', () => {
         draftReferral.complexityLevelId = '110f2405-d944-4c15-836c-0c6684e2aa78'
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:41]: To decrease dependence on the exact values in the factory:

```suggestion
        draftReferral.complexityLevelId = serviceCategory.complexityLevels[1].id
```

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 6]
```diff
@@ -80,7 +58,7 @@ describe('ComplexityLevelPresenter', () => {
     describe('when the referral already has a selected complexity level and there is user input data', () => {
       it('sets checked to true for the complexity level that the user chose', () => {
         draftReferral.complexityLevelId = 'd0db50b0-4a50-4fc7-a006-9c97530e38b2'
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:60]: To decrease dependence on the exact values in the factory:

```suggestion
        draftReferral.complexityLevelId = serviceCategory.complexityLevels[0].id
```

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 7]
```diff
@@ -61,15 +39,15 @@ describe('ComplexityLevelPresenter', () => {
...
 
         expect(presenter.complexityDescriptions.map(description => description.checked)).toEqual([false, true, false])
       })
     })
 
     describe('when there is user input data', () => {
       it('sets checked to true for the complexity level that the user chose', () => {
-        const presenter = new ComplexityLevelPresenter(draftReferral, socialInclusionComplexityLevels, null, {
+        const presenter = new ComplexityLevelPresenter(draftReferral, serviceCategory, null, {
           'complexity-level': '110f2405-d944-4c15-836c-0c6684e2aa78',
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:51]: To decrease dependence on the exact values in the factory

```suggestion
          'complexity-level': serviceCategory.complexityLevels[1],
```

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 8]
```diff
@@ -80,7 +58,7 @@ describe('ComplexityLevelPresenter', () => {
     describe('when the referral already has a selected complexity level and there is user input data', () => {
       it('sets checked to true for the complexity level that the user chose', () => {
         draftReferral.complexityLevelId = 'd0db50b0-4a50-4fc7-a006-9c97530e38b2'
-        const presenter = new ComplexityLevelPresenter(draftReferral, socialInclusionComplexityLevels, null, {
+        const presenter = new ComplexityLevelPresenter(draftReferral, serviceCategory, null, {
           'complexity-level': '110f2405-d944-4c15-836c-0c6684e2aa78',
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:62]: To decrease dependence on the exact values in the factory

```suggestion
          'complexity-level': serviceCategory.complexityLevels[1],
```

**Review by lawrence-forooghian — CHANGES_REQUESTED** [review]

File: server/routes/referrals/complexityLevelPresenter.test.ts [hunk 9]
```diff
@@ -79,9 +57,9 @@ describe('ComplexityLevelPresenter', () => {
 
     describe('when the referral already has a selected complexity level and there is user input data', () => {
       it('sets checked to true for the complexity level that the user chose', () => {
-        draftReferral.complexityLevelId = 'd0db50b0-4a50-4fc7-a006-9c97530e38b2'
-        const presenter = new ComplexityLevelPresenter(draftReferral, socialInclusionComplexityLevels, null, {
-          'complexity-level': '110f2405-d944-4c15-836c-0c6684e2aa78',
+        draftReferral.complexityLevelId = serviceCategory.complexityLevels[1].id
```
Comment (lawrence-forooghian) [inline server/routes/referrals/complexityLevelPresenter.test.ts:60]: I think this should be 0, not 1, to preserve the behaviour of the test.

**Review by lawrence-forooghian — CHANGES_REQUESTED** [review]

**Review by lawrence-forooghian — APPROVED** [review]
